### PR TITLE
[WIP] Implemented opaque resource discovery.

### PR DESCRIPTION
**Pull request by ConnorDoyle**
This is just a sketch of the feature so far.

It's missing:

- [ ] Alignment w/ resource management roadmap
- [ ] Testing
- [ ] Docs

### CODE REVIEW

File: sources.go [hunk 1]
```diff
@@ -7,15 +7,19 @@ import (
...
 )
 
 // FeatureSource represents a source of discovered node features.
 type FeatureSource interface {
 	// Returns a friendly name for this source of node features.
 	Name() string
 
-	// Returns discovered features for this node.
+	// Returns discovered binary features for this node.
 	Discover() ([]string, error)
```
Comment (balajismaniam) [inline sources.go]: Can we make this something more explicit to `DiscoverBinaryResources`?

File: sources.go [hunk 2]
```diff
@@ -7,15 +7,19 @@ import (
...
+
+	// Returns quantities of discovered opaque integer resources for this node.
+	DiscoverResources() (api.ResourceList, error)
```
Comment (balajismaniam) [inline sources.go]: `DiscoverResources()` -> `DiscoverIntegerResources`?

File: sources.go (hunk 1 above)
Comment (balajismaniam) [inline sources.go]: Will have to change this in other places if we go with this change.

File: main.go [hunk 3]
```diff
@@ -73,19 +84,28 @@ func main() {
 	usage := fmt.Sprintf(`%s.
 
   Usage:
-  %s [--no-publish --sources=<sources> --label-whitelist=<pattern>]
+  %s [--sources=<sources>] [--resource-types=<rtypes>]
+	    [--label-whitelist=<pattern>] [--resource-whitelist=<pattern>]
+      [--no-publish]
```
Comment (balajismaniam) [inline main.go]: The indentation seems off here. Does the output look good on your screen?

File: main.go [hunk 4]
```diff
@@ -136,23 +179,48 @@ func main() {
...
 	for _, source := range sources {
-		labelsFromSource, err := getFeatureLabels(source)
-		if err != nil {
-			stderrLogger.Printf("discovery failed for source [%s]: %s", source.Name(), err.Error())
-			stderrLogger.Printf("continuing ...")
-			continue
+
+		var labelsFromSource Labels
+		if _, enabled := enabledResourceTypes[TypeBinary]; enabled {
+			labelsFromSource, err = getFeatureLabels(source)
```
Comment (balajismaniam) [inline main.go]: May be we can change `getFeatureLabels()` -> `getBinaryResources()`. Elsewhere we can document that binary resources are published as labels and integer resources are advertised by patching the capacity of the cluster. Just a thought. LMK WDYT?

File: main.go [hunk 5]
```diff
@@ -136,23 +179,48 @@ func main() {
...
+			if err != nil {
+				stderrLogger.Printf("label discovery failed for source [%s]: %s", source.Name(), err.Error())
+				stderrLogger.Printf("continuing ...")
+				continue
+			}
+		}
+
+		var resourcesFromSource api.ResourceList
+		if _, enabled := enabledResourceTypes[TypeInteger]; enabled {
+			resourcesFromSource, err = getResources(source)
```
Comment (balajismaniam) [inline main.go]: `getResources` -> `getIntegerResources`. See above for my thoughts on this.

File: main.go [hunk 6]
```diff
@@ -162,6 +230,11 @@ func main() {
 		if err != nil {
 			stderrLogger.Fatalf("failed to advertise labels: %s", err.Error())
 		}
+
+		err = advertiseResources(helper, resources)
```
Comment (balajismaniam) [inline main.go]: `advertiseResources()` -> `advertiseIntegerResources()`. Also `advertiseFeatureLabels()` -> `advertiseBinaryResources()`.

File: sources.go (hunk 1 above)
Comment (ConnorDoyle) [inline sources.go]: Yep, good idea.

File: main.go (hunk 3 above)
Comment (ConnorDoyle) [inline main.go]: Looks like the go plugin for my editor may have snuck in some tabs, will fix

File: main.go [hunk 7]
```diff
@@ -217,6 +333,44 @@ func advertiseFeatureLabels(helper APIHelpers, labels Labels) error {
...
+			"op":    "add",
+			"path":  fmt.Sprintf("/status/capacity/%s", escapeForJSONPatch(name)),
+			"value": quantity.String(),
+		})
+	}
+	patchJSON, err := json.Marshal(patch)
+	if err != nil {
+		stderrLogger.Printf("could not prepare patch to advertise opaque resources: %s", err)
+	}
+	return cli.Core().RESTClient().Patch(apitypes.JSONPatchType).Resource("nodes").Name(node.Name).SubResource("status").Body(patchJSON).Do().Error()
```
Comment (balajismaniam) [inline main.go]: Question: what happens when patchJSON is empty (i.e., no integer resources were discovered)?

**Comment by balajismaniam** [comment]
Add tests for `getResources()` and `advertiseResources()`.

**Comment by ConnorDoyle** [comment]
Closing. Opaque Integer Resources are deprecated in v1.8 and removed in v1.9.

**Pull request closed without merging**